src/realm/object-store/sync/sync_manager.cpp

```diff
@@ -746,3 +746,24 @@ util::Optional<SyncAppMetadata> SyncManager::app_metadata() const
     }
     return m_metadata_manager->get_app_metadata();
 }
+
+void SyncManager::close_all_sessions()
+{
+    // log_out will call unregister_session, which tries to acquires m_session_mutex,
+    // so we need to take a copy of the sessions and release the lock before proceeding
+    auto current_sessions = [&] {
+        util::CheckedLockGuard lk(m_session_mutex);
+        std::vector<std::shared_ptr<SyncSession>> current_sessions;
+        std::transform(m_sessions.begin(), m_sessions.end(), std::back_inserter(current_sessions),
+                       [](const auto& session_kv) {
+                           return session_kv.second;
+                       });
+        return current_sessions;
+    }();
+
+    for (auto& session : current_sessions) {
+        session->log_out();
+    }
```

Comment:
Rather than copying the list of sections into a new vector, you can just move the entire map out of `m_sessions`:

```suggestion
    // log_out() will call unregister_session(), which requires m_session_mutex,
    // so we need to iterate over them without holding the lock.
    decltype(m_sessions) sessions;
    {
        util::CheckedLockGuard lk(m_session_mutex);
        m_sessions.swap(sessions);
    }

    for (auto& [_, session] : sessions) {
        session->log_out();
    }
```

This is an irrelevant perf improvement in addition to being less code.

If you're wondering "why not `auto sessions = [&] { util::CheckedLockGuard lk(m_session_mutex); return std::move(m_sessions); }();`?", it's because moving from a standard library type technically leaves it in an unspecified state, even though there's only one possible state for it to be in that satisfies all of the other requirements.

Reply:
Ah cool, I just copied this bit from https://github.com/realm/realm-core/blob/master/src/realm/object-store/sync/sync_manager.cpp#L489-L499, thanks for the improvement!